File: src/aistudio_client/token_factory.py
"""The only permitted attestation boundary for the staging client."""

from __future__ import annotations

import json
from dataclasses import dataclass, field

from .auth import AuthContext, generate_authorization_header
from .http import HttpClient
from .models import RuntimeConfig


@dataclass
class TokenSnapshot:
    token: str
    cookie_records: list[dict[str, str]] = field(default_factory=list)
    transport_profile: dict[str, str] | None = None
    runtime_config: dict | None = None
    logging_context_extension: str | None = None


class StagingTokenFactory:
    def __init__(
        self,
        http: HttpClient,
        url: str,
        waa_api_key: str,
        auth: AuthContext,
        runtime: RuntimeConfig,
        browser_id: str | None = None,
    ) -> None:
        if not url or not waa_api_key:
            raise ValueError("token factory requires URL and opaque.waa_api_key")
        self.http, self.url, self.waa_api_key, self.auth, self.runtime = http, url, waa_api_key, auth, runtime
        self.browser_id = browser_id

    def update_session(self, auth: AuthContext, runtime: RuntimeConfig) -> None:
        self.auth, self.runtime = auth, runtime

    def snapshot(self, digest: str, generate_request: dict) -> TokenSnapshot:
        if len(digest) != 64 or any(char not in "0123456789abcdef" for char in digest):
            raise ValueError("content digest must be lowercase SHA-256 hex")
        authorization = generate_authorization_header(self.auth)
        if not authorization:
            raise ValueError("No session authorization is available for token factory")
        body = {
            "digest": digest, "cookies": self.auth.cookie_header, "authorization": authorization,
            "waaApiKey": self.waa_api_key, "visitId": self.runtime.visit_id,
            "authUser": self.runtime.auth_user, "attestationEnabled": self.runtime.attestation_enabled,
            "generateRequest": generate_request,
            **({"browserId": self.browser_id} if self.browser_id else {}),
        }
        encoded_body = json.dumps(body, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
        response = self.http.request("POST", self.url, headers={"Content-Type": "application/json"}, data=encoded_body)
        if not response.ok:
            from .errors import ClientError
            raise ClientError(f"Token factory failed with HTTP {response.status_code}", phase="ATTESTATION", status=response.status_code, response_body=response.text)
        result = response.json()
        token = result.get("token")
        if not isinstance(token, str) or not token:
            raise ValueError("Token factory response does not contain a token")
        return TokenSnapshot(token, result.get("cookieRecords") or [], result.get("transportProfile"), result.get("runtimeConfig"), result.get("loggingContextExtension"))

File: go/aistudio/token_factory.go
package aistudio

import (
	"context"
	"encoding/json"
	"fmt"
)

type TokenSnapshot struct {
	Token                   string            `json:"token"`
	CookieRecords           []CookieRecord    `json:"cookieRecords"`
	TransportProfile        map[string]string `json:"transportProfile"`
	RuntimeConfig           *RuntimeConfig    `json:"runtimeConfig"`
	LoggingContextExtension string            `json:"loggingContextExtension"`
}
type StagingTokenFactory struct {
	HTTP                      *HTTPClient
	URL, WAAAPIKey, BrowserID string
	Auth                      *AuthContext
	Runtime                   RuntimeConfig
}

func (f *StagingTokenFactory) Snapshot(ctx context.Context, digest string, generateRequest map[string]any) (TokenSnapshot, error) {
	if !regexpDigest(digest) {
		return TokenSnapshot{}, fmt.Errorf("content digest must be lowercase SHA-256 hex")
	}
	authorization := f.Auth.Authorization()
	if authorization == "" {
		return TokenSnapshot{}, fmt.Errorf("No session authorization is available for token factory")
	}
	body, _ := json.Marshal(map[string]any{"digest": digest, "cookies": f.Auth.CookieHeader, "authorization": authorization, "waaApiKey": f.WAAAPIKey, "visitId": f.Runtime.VisitID, "authUser": f.Runtime.AuthUser, "attestationEnabled": f.Runtime.AttestationEnabled, "generateRequest": generateRequest, "browserId": f.BrowserID})
	response, err := f.HTTP.Request(ctx, "POST", f.URL, map[string]string{"Content-Type": "application/json"}, body)
	if err != nil {
		return TokenSnapshot{}, err
	}
	defer response.Body.Close()
	text, err := ReadBody(response)
	if err != nil {
		return TokenSnapshot{}, err
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return TokenSnapshot{}, &ClientError{Message: fmt.Sprintf("Token factory failed with HTTP %d", response.StatusCode), Phase: "ATTESTATION", Status: response.StatusCode, ResponseBody: text}
	}
	var snapshot TokenSnapshot
	if err := json.Unmarshal([]byte(text), &snapshot); err != nil {
		return TokenSnapshot{}, err
	}
	if snapshot.Token == "" {
		return TokenSnapshot{}, fmt.Errorf("Token factory response does not contain a token")
	}
	return snapshot, nil
}
func regexpDigest(value string) bool {
	if len(value) != 64 {
		return false
	}
	for _, r := range value {
		if !(r >= '0' && r <= '9' || r >= 'a' && r <= 'f') {
			return false
		}
	}
	return true
}
